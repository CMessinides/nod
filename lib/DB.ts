import { Pool } from "pg";
import { connectionString } from "../config/db.config";
import { dev } from "../config/server.config";

const pool = new Pool({ connectionString, ssl: !dev });

export default {
	query: pool.query
};
