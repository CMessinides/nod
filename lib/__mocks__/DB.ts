const DB = jest.genMockFromModule("../DB").default;

export function mockRows(rows = [], { once = true } = {}): void {
	const result = { rows, rowCount: rows.length };
	if (once) {
		DB.query.mockResolvedValueOnce(result);
	} else {
		DB.query.mockResolvedValue(result);
	}
}

DB.__mockRows = mockRows;
DB.__mockRows([], { once: false });

export default DB;
