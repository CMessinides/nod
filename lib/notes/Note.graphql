type Note {
	id: ID!
	title: String!
	slug: String!
	createdAt: Date!
	modifiedAt: Date!
	notebook: Notebook!
	content: [NoteChunk]!
}
