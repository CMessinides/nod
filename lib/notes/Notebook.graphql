type Notebook {
	id: ID!
	title: String!
	slug: String!
	description: String
	createdAt: Date!
	notes: [Note]!
}
