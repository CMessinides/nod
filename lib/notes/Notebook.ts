import { NotebookModel } from "./interfaces";
import DB from "../DB";
import pipe from "lodash/fp/pipe";
import { makeSluggable } from "../slugs";
import Note from "./Note";
import sortBy from "lodash/fp/sortBy";
import reverse from "lodash/fp/reverse";

interface NotebookRecord {
	id: number;
	title: string;
	description: string;
	created_at: Date;
}

type NotebookParams = Omit<NotebookModel, "notes" | "slug">;

const prepareNotebook: (notebook: Notebook) => Notebook = pipe(makeSluggable);
const sortByCreatedDate = pipe(
	sortBy<Notebook>("createdAt"),
	reverse
);

export default class Notebook implements NotebookModel {
	id: number;
	title: string;
	slug: string;
	description: string;
	createdAt: Date;

	static async findById(id: number): Promise<Notebook> {
		const { rows, rowCount } = await DB.query(
			"SELECT * FROM notebooks WHERE id = $1 LIMIT 1",
			[id]
		);

		if (rowCount === 0) {
			return null;
		}

		return Notebook.fromRecord(rows[0] as NotebookRecord);
	}

	static async all(): Promise<Notebook[]> {
		const { rows }: { rows: NotebookRecord[] } = await DB.query(
			"SELECT * FROM notebooks;"
		);

		return sortByCreatedDate(rows.map(Notebook.fromRecord));
	}

	static fromRecord({ created_at, ...props }: NotebookRecord): Notebook {
		return new Notebook({ ...props, createdAt: created_at });
	}

	constructor(params: NotebookParams) {
		Object.assign(this, params);
		prepareNotebook(this);
	}

	notes(): Promise<Note[]> {
		return Note.getAllInNotebook(this.id);
	}
}
