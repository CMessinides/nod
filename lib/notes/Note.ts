import { NoteModel, NoteChunkType } from "./interfaces";
import DB from "../DB";
import { makeSluggable } from "../slugs";
import pipe from "lodash/fp/pipe";
import sortBy from "lodash/fp/sortBy";
import reverse from "lodash/fp/reverse";
import Notebook from "./Notebook";
import LinkedList from "../LinkedList";
import NoteTaskListChunk, {
	NoteTaskListChunkRecord
} from "./NoteTaskListChunk";
import NoteTextContentChunk, {
	NoteTextContentChunkRecord
} from "./NoteTextContentChunk";
import { NoteChunkRecord } from "./NoteContentChunk";

interface NoteRecord {
	id: number;
	title: string;
	created_at: Date;
	modified_at: Date;
	notebook_id: number;
}

type NoteParams = Omit<NoteModel, "notebook" | "content" | "slug"> & {
	notebookId: number;
};

const prepareNote: (note: Note) => Note = pipe(makeSluggable);
const sortByModifiedDate = pipe(
	sortBy<Note>("modifiedAt"),
	reverse
);

type NoteContentChunk = NoteTaskListChunk | NoteTextContentChunk;
function modelChunkByType(record: NoteChunkRecord): NoteContentChunk {
	switch (record.type) {
		case NoteChunkType.TASK_LIST:
			return NoteTaskListChunk.fromRecord(record as NoteTaskListChunkRecord);
		case NoteChunkType.TEXT_CONTENT:
			return NoteTextContentChunk.fromRecord(
				record as NoteTextContentChunkRecord
			);
		default:
			throw new Error("Unrecognized content chunk type: " + record.type);
	}
}

export default class Note implements NoteModel {
	id: number;
	title: string;
	slug: string;
	createdAt: Date;
	modifiedAt: Date;
	private notebookId: number;

	static async findById(id: number): Promise<Note> {
		const { rows, rowCount } = await DB.query(
			"SELECT * FROM notes WHERE id = $1 LIMIT 1",
			[id]
		);

		if (rowCount === 0) {
			return null;
		}

		return Note.fromRecord(rows[0] as NoteRecord);
	}

	static async getAllInNotebook(notebookId: number): Promise<Note[]> {
		const { rows }: { rows: NoteRecord[] } = await DB.query(
			"SELECT * FROM notes WHERE notebook_id = $1",
			[notebookId]
		);

		return sortByModifiedDate(rows.map(Note.fromRecord));
	}

	static fromRecord({
		id,
		title,
		created_at,
		modified_at,
		notebook_id
	}: NoteRecord): Note {
		return new Note({
			id,
			title,
			createdAt: created_at,
			modifiedAt: modified_at,
			notebookId: notebook_id
		});
	}

	constructor(params: NoteParams) {
		for (const key of Object.keys(params)) {
			this[key] = params[key];
		}

		prepareNote(this);
	}

	notebook(): Promise<Notebook> {
		return Notebook.findById(this.notebookId);
	}

	async content(): Promise<NoteContentChunk[]> {
		const { rows }: { rows: NoteChunkRecord[] } = await DB.query(
			"SELECT note_chunks.id, prev_chunk_id, type, name, list_id, text FROM note_chunks FULL JOIN note_text_content_chunks ON note_chunks.id = note_text_content_chunks.id FULL JOIN (SELECT note_task_list_chunks.id, list_id, name FROM note_task_list_chunks JOIN task_lists ON note_task_list_chunks.list_id = task_lists.id) AS full_note_task_list_chunks ON note_chunks.id = full_note_task_list_chunks.id WHERE note_id = $1",
			[this.id]
		);

		return Array.from(LinkedList(rows, "id", "prev_chunk_id")).map(
			modelChunkByType
		);
	}
}
