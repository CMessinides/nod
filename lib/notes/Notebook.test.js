import Notebook from "./Notebook";

jest.mock("../DB");
import DB from "../DB";
jest.mock("./Note");
import Note from "./Note";
import MockRecords from "./fixtures/NotebookRecords";

describe("static methods", () => {
	describe("findById", () => {
		it("should return the notebook with the given ID", () => {
			const record = MockRecords[0];
			const { id } = record;
			DB.__mockRows([record]);

			expect(Notebook.findById(id)).resolves.toMatchSnapshot();
		});
	});
});

describe("instance methods", () => {
	describe("notes", () => {
		it("should return the notes in the notebook", () => {
			const notebook = Notebook.fromRecord(MockRecords[0]);
			const notes = [
				new Note({ id: 1, notebookId: notebook.id }),
				new Note({ id: 2, notebookId: notebook.id })
			];
			Note.getAllInNotebook.mockResolvedValueOnce(notes);

			expect(notebook.notes()).resolves.toBe(notes);
		});
	});
});
