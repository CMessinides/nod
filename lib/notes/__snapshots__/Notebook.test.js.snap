// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`static methods all should return all notebooks 1`] = `
Array [
  Notebook {
    "createdAt": 1970-01-18T16:01:01.001Z,
    "description": "Mauris enim leo, rhoncus sed, vestibulum sit amet, cursus id, turpis.",
    "id": 2,
    "slug": "geochelone-radiata--2",
    "title": "Geochelone radiata",
  },
  Notebook {
    "createdAt": 1970-01-18T15:21:16.522Z,
    "description": "Nunc nisl. Duis bibendum, felis sed interdum venenatis, turpis enim blandit mi, in porttitor pede justo eu massa.",
    "id": 3,
    "slug": "snycerus-caffer--3",
    "title": "Snycerus caffer",
  },
  Notebook {
    "createdAt": 1970-01-18T12:45:05.294Z,
    "description": "Mauris lacinia sapien quis libero. Nullam sit amet turpis elementum ligula vehicula consequat. Morbi a ipsum.",
    "id": 4,
    "slug": "anas-punctata--4",
    "title": "Anas punctata",
  },
  Notebook {
    "createdAt": 1970-01-18T10:34:31.205Z,
    "description": "In hac habitasse platea dictumst. Morbi vestibulum, velit id pretium iaculis, diam erat fermentum justo, nec condimentum neque sapien placerat ante. Nulla justo.",
    "id": 1,
    "slug": "dendrocitta-vagabunda--1",
    "title": "Dendrocitta vagabunda",
  },
]
`;

exports[`static methods findById should return the notebook with the given ID 1`] = `
Notebook {
  "createdAt": 1970-01-18T10:34:31.205Z,
  "description": "In hac habitasse platea dictumst. Morbi vestibulum, velit id pretium iaculis, diam erat fermentum justo, nec condimentum neque sapien placerat ante. Nulla justo.",
  "id": 1,
  "slug": "dendrocitta-vagabunda--1",
  "title": "Dendrocitta vagabunda",
}
`;
